drivers/bluetooth/hci/spi.c

```diff
@@ -186,42 +182,88 @@ static void release_cs(void)
 {
 	gpio_pin_set_dt(&bus.config.cs.gpio, 0);
 }
-
-static bool irq_pin_high(void)
+static int bt_spi_get_header(uint8_t op, uint8_t *size)
 {
-	int pin_state;
-
-	pin_state = gpio_pin_get_dt(&irq_gpio);
-
-	LOG_DBG("IRQ Pin: %d", pin_state);
+	uint8_t header_master[5] = {op, 0, 0, 0, 0};
+	uint8_t header_slave[5];
+	uint8_t size_offset, attempts;
+	int ret;
 
-	return pin_state > 0;
-}
+	if (op == SPI_READ) {
+		if (!IS_IRQ_HIGH) {
+			*size = 0;
+			return 0;
+		}
+		size_offset = STATUS_HEADER_TOREAD;
+	} else if (op == SPI_WRITE) {
+		size_offset = STATUS_HEADER_TOWRITE;
+	}
 
-static void init_irq_high_loop(void)
-{
 	attempts = IRQ_HIGH_MAX_READ;
-}
-
-static bool exit_irq_high_loop(void)
-{
-	/* Limit attempts on BlueNRG-MS as we might */
-	/* enter this loop with nothing to read */
+	do {
+		if (op == SPI_READ) {
+			/* Keep checking that IRQ is still high, if we need to read */
+			if (!IS_IRQ_HIGH) {
+				*size = 0;
+				return 0;
+			}
+		}
+		assert_cs();
+		ret = bt_spi_transceive(header_master, 5, header_slave, 5);
+		if (ret) {
+			/* SPI transaction failed */
+			break;
+		}
 
-	attempts--;
+		*size = (header_slave[STATUS_HEADER_READY] == READY_NOW) ?
+				header_slave[size_offset] : 0;
+		attempts--;
+	} while ((*size == 0) && attempts);
 
-	return attempts;
+	return ret;
 }
 
+/* Other Boards */
 #else
 
-#define kick_cs(...)
 #define release_cs(...)
-#define irq_pin_high(...) 0
-#define init_irq_high_loop(...)
-#define exit_irq_high_loop(...) 1
+#define READ_CONDITION false
+#define WRITE_DATA_CONDITION(ret, rx_first) (rx_first != 0U || ret)
+
+static int bt_spi_get_header(uint8_t op, uint8_t *size)
```

Comment:
have you tested also the "else" side of the code?

Reply:
Yes. It has been tested by 96b_carbon board which uses STM32F401RET and nRF51822 as main MCU and BLE module respectively.